### PR TITLE
Handle branch names differing only by case

### PR DESCRIPTION
**Pull request by hdgarrood**
### Description of the change

Escape upper case characters in branch names to ensure that branches
which only differ in case don't cause issues on case-insensitive file
systems.

Also switch to using an allowlist for escaping characters: it seems a
little safer to specify the characters which don't need escaping (and to
escape everything else), than to specify characters which do need
escaping.

Finally, add a test that `getCacheVersionDir` is injective.

### Checklist:

- [x] Added the change to the "Unreleased" section of the changelog
- [ ] ~~Added some example of the new feature to the `README`~~
- [x] Added a test for the contribution (if applicable)

**P.S.**: the above checks are not compulsory to get a change merged, so you may skip them. However, taking care of them will result in less work for the maintainers and will be much appreciated 😊

### CODE REVIEW

**Comment by hdgarrood** [comment]
It turns out that using QuickCheck to check that a function is injective requires a bit more effort than I thought it would. This is the simplest and most sensible thing I could come up with which reliably fails if I modify `getVersionCacheDir` to no longer be injective.

File: package.yaml [hunk 1]
```diff
@@ -134,3 +134,7 @@ tests:
     - directory
     - temporary
     - extra
+    - containers
+    - QuickCheck
+    - spago
+    - extra
```
Comment (f-f) [inline package.yaml]:
```suggestion
```

Comment (hdgarrood) [reply package.yaml]: I don't think this can be removed, it's being used for `nubOrd`

Comment (f-f) [reply package.yaml]: It's already being imported in the line preceding your additions

Comment (hdgarrood) [reply package.yaml]: Oh of course, haha :man_facepalming:

**Pull request merged**